File: 00_Alternate_Languages/33_Dice/Julia/Dice.jl
#=
Port of Dice from BASIC Computer Games (1978)

This "game" simulates a given number of dice rolls, and returns the 
count for each possible total.

The only change that has been made from the original program, is that 
when asking if the user wants to play again, any string starting with
y or Y will be accepted, instead of only YES.
=#

function main()
    # Array to store the counts for each total.
    # There are 11 possible totals.
    counts = [0 for i in 1:11]

    # Print intro text
    println("\n                   Dice")
    println("Creative Computing  Morristown, New Jersey")
    println("\n\n")
    println("This program simulates the rolling of a")
    println("pair of dice.")
    println("You enter the number of times you want the computer to")
    println("'roll' the dice.   Watch out, very large numbers take")
    println("a long time.  In particular, numbers over 5000.")

    still_playing = true
    while still_playing
        println()
        print("How many rolls? ")

        # Get user input for number of dice rolls
        rolls = readline()
        rolls = parse(Int64, rolls)

        # Roll dice the specified number of times and update total count
        for _ in 1:rolls
            dice_roll = rand(1:6, 2)
            dice_sum = sum(dice_roll)

            # The index is one less than the sum, as a sum of 1 is impossible,
            # the array will only have 11 values
            counts[dice_sum-1] += 1
        end

        # Display results
        println("\nTotal Spots   Number of Times")
        for i in 1:8
            print(" ")
            print(i+1)
            print("             ")
            println(counts[i])
        end
        for i in 9:11
            print(" ")
            print(i+1)
            print("            ")
            println(counts[i])
        end

        # Ask try again
        print("\nTry Again? ")
        input = readline()
        if length(input) > 0 && uppercase(input)[1] == 'Y'
             # If game is continued, resets total counts
            counts = [0 for i in 1:11]
        else
            still_playing = false
        end
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: 00_Alternate_Languages/33_Dice/nim/dice.nim
import std/[random,strutils]

var
  a,b,r,x: int
  f: array[2..12, int]
  z: string
  retry: bool = true

echo spaces(34), "DICE"
echo spaces(15), "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY"
echo "\n"
echo "THIS PROGRAM SIMULATES THE ROLLING OF A PAIR OF DICE."
echo "YOU ENTER THE NUMBER OF TIMES YOU WANT THE COMPUTER TO"
echo "'ROLL' THE DICE.  WATCH OUT, VERY LARGE NUMBERS TAKE"
echo "A LONG TIME.  IN PARTICULAR, NUMBERS OVER 5000."

while(retry):
  echo "\n"
  echo "HOW MANY ROLLS"
  x = readLine(stdin).parseInt()
  for v in 2..12:
    f[v] = 0 # Initialize array to 0
  for s in 1..x:
    a = rand(1..6) # Die 1
    b = rand(1..6) # Die 2
    r = a + b # Sum of dice
    f[r] += 1 # Increment array count of dice sum result
  echo ""
  echo "TOTAL SPOTS: ", "NUMBER OF TIMES"
  for v in 2..12:
    echo v, ": ", f[v] # Print out counts for each possible result
  echo "\n"
  echo "TRY AGAIN?"
  z = readLine(stdin).normalize()
  retry = (z=="yes") or (z=="y")
